File: Assets/Scripts/Core/GameSession.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;
using Bunker.Content;
using Bunker.Core.Effects;

namespace Bunker.Core
{
    // The authoritative brain of a single game session.
    // Pure C# (no MonoBehaviour) — UI and networking subscribe to its events
    // and call its public methods, never touch player data directly.
    public class GameSession
    {
        public List<PlayerData> Players { get; private set; }
        public int CurrentRound { get; private set; }
        public GamePhase Phase { get; private set; } = GamePhase.Lobby;

        private readonly GameSessionConfig _config;
        private readonly CharacterCardGenerator _cardGenerator;
        private readonly TurnOrderService _turnOrder = new();
        private readonly SpecialCardEffectRegistry _effectRegistry = new();
        private readonly Dictionary<string, string> _currentVotes = new(); // voterId -> targetId
        private List<string> _tiebreakerCandidateIds; // null when not in a tiebreaker
        private int _tiebreakerAttempts;
        private const int MaxTiebreakerAttempts = 1; // one re-vote, then give up


        // Fired whenever the phase transitions. UI listens to this to switch screens.
        public event Action<GamePhase> OnPhaseChanged;

        // Fired when a specific player is eliminated at the end of a round.
        public event Action<PlayerData> OnPlayerEliminated;

        // Fired once the win condition is reached, with the final survivors.
        public event Action<List<PlayerData>> OnGameOver;

        // Fired when a round officially starts (useful for UI to reset round-scoped state).
        public event Action<int> OnRoundStarted;

        // Fired whenever a player reveals one of their traits.
        // UI uses this to update that player's visible card.
        public event Action<PlayerData, CharacterTrait> OnTraitRevealed;

        // Fired when every active player has had a turn this round
        // (a full pass through the reveal order is complete).
        public event Action OnRevealPassCompleted;

        // Fired when a special card is successfully used, for UI feedback
        // (e.g. "Player X used their special card").
        public event Action<PlayerData, SpecialCard> OnSpecialCardUsed;

        // Fired when the discussion timer should start; the actual countdown
        // is owned by the caller (UI/network), GameSession only marks the phase.
        public event Action<int> OnDiscussionStarted; // duration in seconds, chosen by caller

        // Fired with the full result once a round's voting is resolved.
        public event Action<VotingResult> OnVotingResolved;

        // Fired once the game ends, with the full reason + survivor list.
        // Replaces the simpler List<PlayerData>-only event from section 1.3.
        public event Action<GameOverResult> OnGameOverResolved;

        public string CurrentTurnPlayerId => _turnOrder.CurrentPlayerId;

        public GameSession(
            List<PlayerData> players,
            GameSessionConfig config,
            CharacterCardGenerator cardGenerator)
        {
            if (players == null || players.Count == 0)
                throw new ArgumentException("GameSession requires at least one player.");

            Players = players;
            _config = config;
            _cardGenerator = cardGenerator;
        }

        // --- Pre-start validation --------------------------------------------

        // Called by the lobby UI before enabling the "Start" button, and again
        // defensively inside StartGame itself.
        public GameStartValidationResult ValidateCanStart()
        {
            if (Phase != GamePhase.Lobby)
                return GameStartValidationResult.Fail("Game has already started.");

            if (Players.Count == 0)
                return GameStartValidationResult.Fail("No players in the lobby.");

            if (_config.SurvivorsTarget <= 0)
                return GameStartValidationResult.Fail("Survivors target must be at least 1.");

            // Need strictly more players than the target, otherwise there is
            // nothing to play — the game would already be "won" at round 0.
            if (Players.Count <= _config.SurvivorsTarget)
            {
                return GameStartValidationResult.Fail(
                    $"Need more than {_config.SurvivorsTarget} players to start " +
                    $"(currently {Players.Count}).");
            }

            // Content sanity check: every player must have a full set of traits
            // ready to be dealt. Catches a misconfigured/empty trait pool early
            // instead of failing deep inside CharacterCardGenerator.
            if (!_cardGenerator.HasEnoughContentFor(Players.Count))
            {
                return GameStartValidationResult.Fail("Not enough card content configured for this many players.");
            }

            return GameStartValidationResult.Ok();
        }

        public void StartGame()
        {
            var validation = ValidateCanStart();
            if (!validation.CanStart)
            {
                Debug.LogError($"[GameSession] Cannot start game: {validation.FailReason}");
                return;
            }

            SetPhase(GamePhase.Dealing);
            _cardGenerator.DealToPlayers(Players);
            _turnOrder.Initialize(Players);

            CurrentRound = 0;
            StartNextRound();
        }

        // --- Lifecycle -----------------------------------------------------

        private void StartNextRound()
        {
            CurrentRound++;

            foreach (var player in ActivePlayers())
            {
                player.ResetPerRoundFlags();
            }

            _currentVotes.Clear();
            _turnOrder.RebuildForRound(CurrentRound, ActivePlayers());

            SetPhase(GamePhase.Reveal);
            OnRoundStarted?.Invoke(CurrentRound);
        }

        // --- Reveal phase ----------------------------------------------------

        // Attempts to reveal one trait category for the player whose turn it currently is.
        // Returns false (and does nothing) if the request is invalid, so the caller
        // (UI / network RPC handler) can show appropriate feedback instead of crashing.
        public bool RevealNextTrait(string playerId, CardCategory category)
        {
            if (Phase != GamePhase.Reveal)
            {
                Debug.LogWarning("[GameSession] RevealNextTrait called outside of Reveal phase.");
                return false;
            }

            if (!_turnOrder.IsPlayersTurn(playerId))
            {
                Debug.LogWarning($"[GameSession] It is not {playerId}'s turn.");
                return false;
            }

            var player = GetPlayer(playerId);
            if (player == null || player.IsEliminated)
            {
                Debug.LogWarning($"[GameSession] Unknown or eliminated player: {playerId}");
                return false;
            }

            if (player.IsCategoryRevealed(category))
            {
                Debug.LogWarning($"[GameSession] {playerId} already revealed {category}.");
                return false;
            }

            var trait = player.GetTrait(category);
            if (trait == null)
            {
                Debug.LogWarning($"[GameSession] {playerId} has no trait for category {category}.");
                return false;
            }

            player.RevealCategory(category);
            OnTraitRevealed?.Invoke(player, trait);

            AdvanceToNextTurn();
            return true;
        }

        private void AdvanceToNextTurn()
        {
            bool passContinues = _turnOrder.AdvanceTurn();

            if (!passContinues)
            {
                OnRevealPassCompleted?.Invoke();
                // Reveal phase ends after one full pass; the caller (UI/network)
                // decides when to actually transition to Discussion.
                // See note below.
            }
        }

        // Convenience check for the UI/network layer to know if a full reveal
        // pass has finished for all active players this round.
        public bool HasCompletedRevealPass()
        {
            return ActivePlayers().All(p => p.RevealedCategories.Count > 0);
        }

        // --- Phase transitions ---------------------------------------------
        // These are intentionally simple setters; concrete round logic
        // (reveal order, voting rules) lives in later sections (1.4, 1.5, 1.6)
        // and will call into these transitions.

        public void SetPhase(GamePhase newPhase)
        {
            Phase = newPhase;
            OnPhaseChanged?.Invoke(newPhase);
        }

        // --- Discussion phase -------------------------------------------------

        // Called once the reveal pass is done (see OnRevealPassCompleted from 1.4).
        // durationSeconds is passed by the caller since GameSession doesn't own real time.
        public void StartDiscussionPhase(int durationSeconds)
        {
            if (Phase != GamePhase.Reveal)
            {
                Debug.LogWarning("[GameSession] StartDiscussionPhase called outside of Reveal phase.");
                return;
            }

            SetPhase(GamePhase.Discussion);
            OnDiscussionStarted?.Invoke(durationSeconds);
        }

        // Called by the caller's timer once discussion time is up (or all players
        // signal ready — that policy lives outside GameSession).
        public void StartVotingPhase()
        {
            if (Phase != GamePhase.Discussion)
            {
                Debug.LogWarning("[GameSession] StartVotingPhase called outside of Discussion phase.");
                return;
            }

            _tiebreakerCandidateIds = null;
            _tiebreakerAttempts = 0;
            _currentVotes.Clear();
            SetPhase(GamePhase.Voting);
        }

        // --- Voting  ------------------------------------------------------------

        public bool CastVote(string voterPlayerId, string targetPlayerId)
        {
            bool inTiebreaker = Phase == GamePhase.VotingTiebreaker;

            if (Phase != GamePhase.Voting && !inTiebreaker)
            {
                Debug.LogWarning("[GameSession] Vote cast outside of Voting phase, ignored.");
                return false;
            }

            var voter = GetPlayer(voterPlayerId);
            if (voter == null || voter.IsEliminated)
            {
                Debug.LogWarning($"[GameSession] Invalid voter: {voterPlayerId}");
                return false;
            }

            if (voterPlayerId == targetPlayerId)
            {
                Debug.LogWarning("[GameSession] Self-voting is not allowed.");
                return false;
            }

            var target = GetPlayer(targetPlayerId);
            if (target == null || target.IsEliminated)
            {
                Debug.LogWarning($"[GameSession] Invalid vote target: {targetPlayerId}");
                return false;
            }

            if (target.HasVoteImmunityThisRound)
            {
                Debug.LogWarning($"[GameSession] {targetPlayerId} is immune this round, vote rejected.");
                return false;
            }

            // During a tiebreaker, votes may only go to the tied candidates
            // from the previous round of voting.
            if (inTiebreaker && !_tiebreakerCandidateIds.Contains(targetPlayerId))
            {
                Debug.LogWarning("[GameSession] Vote target is not part of the tiebreaker candidates.");
                return false;
            }

            _currentVotes[voterPlayerId] = targetPlayerId;
            return true;
        }

        // Called by the caller once every active player has voted (or a voting
        // timer expired). Resolves the round and transitions phases accordingly.
        public VotingResult ResolveVotes()
        {
            var result = TallyVotes();
            OnVotingResolved?.Invoke(result);

            switch (result.ResultType)
            {
                case VotingResult.Outcome.PlayerEliminated:
                    result.EliminatedPlayer.IsEliminated = true;
                    OnPlayerEliminated?.Invoke(result.EliminatedPlayer);
                    FinishRound();
                    break;

                case VotingResult.Outcome.TieRequiresRevote:
                    _tiebreakerCandidateIds = result.TiedCandidates.Select(p => p.PlayerId).ToList();
                    _tiebreakerAttempts++;
                    _currentVotes.Clear();
                    SetPhase(GamePhase.VotingTiebreaker);
                    break;

                case VotingResult.Outcome.TieUnresolvedNoElimination:
                case VotingResult.Outcome.NoVotesCast:
                    FinishRound();
                    break;
            }

            return result;
        }

        private VotingResult TallyVotes()
        {
            var voteCounts = new Dictionary<string, int>();

            foreach (var targetId in _currentVotes.Values)
            {
                voteCounts.TryGetValue(targetId, out int current);
                voteCounts[targetId] = current + 1;
            }

            if (voteCounts.Count == 0)
            {
                return new VotingResult
                {
                    ResultType = VotingResult.Outcome.NoVotesCast,
                    VoteCounts = voteCounts
                };
            }

            int topCount = voteCounts.Values.Max();
            var topIds = voteCounts.Where(kv => kv.Value == topCount).Select(kv => kv.Key).ToList();

            if (topIds.Count == 1)
            {
                return new VotingResult
                {
                    ResultType = VotingResult.Outcome.PlayerEliminated,
                    EliminatedPlayer = GetPlayer(topIds[0]),
                    VoteCounts = voteCounts
                };
            }

            // Tie: decide whether to allow a re-vote or give up.
            var tiedPlayers = topIds.Select(GetPlayer).ToList();

            bool canRetry = _tiebreakerAttempts < MaxTiebreakerAttempts;

            return new VotingResult
            {
                ResultType = canRetry
                    ? VotingResult.Outcome.TieRequiresRevote
                    : VotingResult.Outcome.TieUnresolvedNoElimination,
                TiedCandidates = tiedPlayers,
                VoteCounts = voteCounts
            };
        }

        // --- Round wrap-up -------------------------------------------------

        private void FinishRound()
        {
            SetPhase(GamePhase.RoundResult);

            if (IsGameOver())
            {
                EndGame();
            }
            else
            {
                StartNextRound();
            }
        }

        // --- Win condition ---------------------------------------------------

        private bool IsGameOver()
        {
            return ActivePlayers().Count() <= _config.SurvivorsTarget;
        }

        private void EndGame()
        {
            SetPhase(GamePhase.GameOver);

            var survivors = ActivePlayers().ToList();

            var reason = survivors.Count > 0
                ? GameOverResult.Reason.SurvivorsTargetReached
                : GameOverResult.Reason.AllPlayersEliminated;

            var result = new GameOverResult
            {
                EndReason = reason,
                Survivors = survivors
            };

            OnGameOverResolved?.Invoke(result);
        }

        // --- Special Cards ---------------------------------------------------

        // General-purpose entry point for effects that don't need extra parameters
        // (VoteImmunity, RevealHiddenTrait, ForceRevealAll).
        public SpecialCardEffectResult UseSpecialCard(string casterPlayerId, string targetPlayerId)
        {
            var caster = GetPlayer(casterPlayerId);
            if (caster == null || caster.IsEliminated)
                return SpecialCardEffectResult.Fail("Invalid caster.");

            if (caster.HasUsedSpecialCard)
                return SpecialCardEffectResult.Fail("Special card already used.");

            if (caster.Special == null)
                return SpecialCardEffectResult.Fail("Player has no special card.");

            if (!_effectRegistry.TryGet(caster.Special.EffectType, out var effect))
                return SpecialCardEffectResult.Fail($"Unknown effect type: {caster.Special.EffectType}");

            var target = targetPlayerId != null ? GetPlayer(targetPlayerId) : null;

            if (!effect.CanApply(this, caster, target))
                return SpecialCardEffectResult.Fail("Effect cannot be applied in this context.");

            // Effect-specific logic that needs extra data lives here, keyed by
            // effect type, rather than bloating ISpecialCardEffect's signature
            // with parameters only some effects need.
            SpecialCardEffectResult result = caster.Special.EffectType switch
            {
                SpecialCardEffectType.RevealHiddenTrait => ApplyRevealHiddenTrait(caster, target),
                SpecialCardEffectType.VoteImmunity => ApplyVoteImmunity(effect, caster, target),
                SpecialCardEffectType.ForceRevealAll => ApplyForceRevealAll(caster, target),
                _ => SpecialCardEffectResult.Fail("Effect requires extra parameters, use the dedicated overload.")
            };

            if (result.Success)
            {
                caster.HasUsedSpecialCard = true;
                OnSpecialCardUsed?.Invoke(caster, caster.Special);
            }

            return result;
        }

        // Dedicated overload for SwapTrait, which needs to know which category to swap.
        public SpecialCardEffectResult UseSwapTraitSpecialCard(
            string casterPlayerId, string targetPlayerId, CardCategory category)
        {
            var caster = GetPlayer(casterPlayerId);
            var target = GetPlayer(targetPlayerId);

            if (caster == null || target == null || caster.IsEliminated || target.IsEliminated)
                return SpecialCardEffectResult.Fail("Invalid caster or target.");

            if (caster.HasUsedSpecialCard)
                return SpecialCardEffectResult.Fail("Special card already used.");

            if (caster.Special?.EffectType != SpecialCardEffectType.SwapTrait)
                return SpecialCardEffectResult.Fail("Player's special card is not a swap effect.");

            if (caster.IsCategoryRevealed(category) || target.IsCategoryRevealed(category))
                return SpecialCardEffectResult.Fail("Category must be hidden for both players.");

            var casterTrait = caster.GetTrait(category);
            var targetTrait = target.GetTrait(category);

            caster.ReplaceTrait(category, targetTrait);
            target.ReplaceTrait(category, casterTrait);

            caster.HasUsedSpecialCard = true;
            OnSpecialCardUsed?.Invoke(caster, caster.Special);

            return SpecialCardEffectResult.Ok();
        }

        private SpecialCardEffectResult ApplyRevealHiddenTrait(PlayerData caster, PlayerData target)
        {
            var hiddenTrait = target.Traits.FirstOrDefault(t => !target.IsCategoryRevealed(t.Category));
            if (hiddenTrait == null)
                return SpecialCardEffectResult.Fail("Target has no hidden traits.");

            // Note: this does NOT call target.RevealCategory() — the target's
            // card stays hidden from everyone else. Only the caster receives
            // this trait in the result, to be shown privately in their UI.
            return SpecialCardEffectResult.Ok(hiddenTrait);
        }

        private SpecialCardEffectResult ApplyVoteImmunity(ISpecialCardEffect effect, PlayerData caster, PlayerData target)
        {
            effect.Apply(this, caster, target);
            return SpecialCardEffectResult.Ok();
        }

        private SpecialCardEffectResult ApplyForceRevealAll(PlayerData caster, PlayerData target)
        {
            var hiddenTrait = target.Traits.FirstOrDefault(t => !target.IsCategoryRevealed(t.Category));
            if (hiddenTrait == null)
                return SpecialCardEffectResult.Fail("Target has no hidden traits.");

            target.RevealCategory(hiddenTrait.Category);
            OnTraitRevealed?.Invoke(target, hiddenTrait); // public — everyone sees it

            return SpecialCardEffectResult.Ok(hiddenTrait);
        }

        // --- Helpers ---------------------------------------------------------

        public IEnumerable<PlayerData> ActivePlayers()
        {
            return Players.Where(p => !p.IsEliminated);
        }

        public PlayerData GetPlayer(string playerId)
        {
            return Players.FirstOrDefault(p => p.PlayerId == playerId);
        }
    }
}

File: bunker-server/src/game/GameSession.ts
import { PlayerData } from "./PlayerData";
import { CharacterTrait } from "./CharacterTrait";
import { CardCategory } from "./CardCategory";
import { SpecialCard, SpecialCardEffectType } from "./SpecialCard";
import { GamePhase } from "./GamePhase";
import { GameSessionConfig } from "./GameSessionConfig";
import { CharacterCardGenerator } from "./content/CharacterCardGenerator";
import { VotingResult, VotingOutcome } from "./VotingResult";
import { GameOverResult, GameOverReason } from "./GameOverResult";
import { GameStartValidationResult, validationOk, validationFail } from "./GameStartValidationResult";
import { TypedEventEmitter } from "./events/TypedEventEmitter";
import { TurnOrderService } from "./TurnOrderService";
import { SpecialCardEffectRegistry } from "./effects/SpecialCardEffectRegistry";
import { ISpecialCardEffect } from "./effects/ISpecialCardEffect";
import { SpecialCardEffectResult, ok, fail } from "./effects/SpecialCardEffectResult";

// Event map for GameSession's TypedEventEmitter. Each entry mirrors one of
// the C# `event Action<T>` fields accumulated across sections 1.3-1.7.
// Declared as a `type` rather than an `interface` — TS's generic constraint
// check (`Events extends Record<string, any[]>` in TypedEventEmitter) only
// accepts type aliases here; a named interface without an explicit index
// signature fails that check even though the shape is identical.
type GameSessionEvents = {
  phaseChanged: [GamePhase];
  roundStarted: [number];
  traitRevealed: [PlayerData, CharacterTrait];
  revealPassCompleted: [];
  specialCardUsed: [PlayerData, SpecialCard];
  discussionStarted: [number]; // duration in seconds
  votingResolved: [VotingResult];
  playerEliminated: [PlayerData];
  gameOverResolved: [GameOverResult];
};

// Direct, fully merged port of the C# GameSession (sections 1.3-1.7 combined
// into their final form). This is the TypeScript equivalent of the complete
// C# class after all incremental additions — turn order (1.4), special card
// effects (1.5), full voting/tiebreaker logic (1.6), and start validation (1.7).
export class GameSession extends TypedEventEmitter<GameSessionEvents> {
  public readonly players: PlayerData[];
  public currentRound = 0;
  public phase: GamePhase = GamePhase.Lobby;

  protected readonly config: GameSessionConfig;
  protected readonly cardGenerator: CharacterCardGenerator;

  private readonly turnOrder = new TurnOrderService();
  private readonly effectRegistry = new SpecialCardEffectRegistry();

  private currentVotes = new Map<string, string>(); // voterId -> targetId
  private tiebreakerCandidateIds: string[] | null = null;
  private tiebreakerAttempts = 0;
  private static readonly MAX_TIEBREAKER_ATTEMPTS = 1; // one re-vote, then give up

  constructor(players: PlayerData[], config: GameSessionConfig, cardGenerator: CharacterCardGenerator) {
    super();

    if (!players || players.length === 0) {
      throw new Error("GameSession requires at least one player.");
    }

    this.players = players;
    this.config = config;
    this.cardGenerator = cardGenerator;
  }

  get currentTurnPlayerId(): string | null {
    return this.turnOrder.currentPlayerId;
  }

  // --- Pre-start validation --------------------------------------------

  // Called by the room/lobby code before allowing "Start Game", and again
  // defensively inside startGame() itself.
  validateCanStart(): GameStartValidationResult {
    if (this.phase !== GamePhase.Lobby) {
      return validationFail("Game has already started.");
    }

    if (this.players.length === 0) {
      return validationFail("No players in the lobby.");
    }

    if (this.config.survivorsTarget <= 0) {
      return validationFail("Survivors target must be at least 1.");
    }

    // Need strictly more players than the target, otherwise there is
    // nothing to play - the game would already be "won" at round 0.
    if (this.players.length <= this.config.survivorsTarget) {
      return validationFail(
        `Need more than ${this.config.survivorsTarget} players to start (currently ${this.players.length}).`
      );
    }

    // Content sanity check: every player must have a full set of traits
    // ready to be dealt. Catches a misconfigured/empty trait pool early
    // instead of failing deep inside CharacterCardGenerator.
    if (!this.cardGenerator.hasEnoughContentFor(this.players.length)) {
      return validationFail("Not enough card content configured for this many players.");
    }

    return validationOk();
  }

  // --- Lifecycle -----------------------------------------------------

  startGame(): void {
    const validation = this.validateCanStart();
    if (!validation.canStart) {
      console.error(`[GameSession] Cannot start game: ${validation.failReason}`);
      return;
    }

    this.setPhase(GamePhase.Dealing);
    this.cardGenerator.dealToPlayers(this.players);
    this.turnOrder.initialize(this.players);

    this.currentRound = 0;
    this.startNextRound();
  }

  protected startNextRound(): void {
    this.currentRound++;

    for (const player of this.activePlayers()) {
      player.resetPerRoundFlags();
    }

    this.currentVotes.clear();
    this.turnOrder.rebuildForRound(this.currentRound, this.activePlayers());

    this.setPhase(GamePhase.Reveal);
    this.emit("roundStarted", this.currentRound);
  }

  // --- Phase transitions ---------------------------------------------

  protected setPhase(newPhase: GamePhase): void {
    this.phase = newPhase;
    this.emit("phaseChanged", newPhase);
  }

  // --- Reveal phase ----------------------------------------------------

  // Attempts to reveal one trait category for the player whose turn it
  // currently is. Returns false (does nothing further) on any invalid
  // request, mirroring the C# RevealNextTrait's bool-return contract.
  revealNextTrait(playerId: string, category: CardCategory): boolean {
    if (this.phase !== GamePhase.Reveal) {
      console.warn("[GameSession] revealNextTrait called outside of Reveal phase.");
      return false;
    }

    if (!this.turnOrder.isPlayersTurn(playerId)) {
      console.warn(`[GameSession] It is not ${playerId}'s turn.`);
      return false;
    }

    const player = this.getPlayer(playerId);
    if (!player || player.isEliminated) {
      console.warn(`[GameSession] Unknown or eliminated player: ${playerId}`);
      return false;
    }

    if (player.isCategoryRevealed(category)) {
      console.warn(`[GameSession] ${playerId} already revealed ${category}.`);
      return false;
    }

    const trait = player.getTrait(category);
    if (!trait) {
      console.warn(`[GameSession] ${playerId} has no trait for category ${category}.`);
      return false;
    }

    player.revealCategory(category);
    this.emit("traitRevealed", player, trait);

    this.advanceToNextTurn();
    return true;
  }

  private advanceToNextTurn(): void {
    const passContinues = this.turnOrder.advanceTurn();

    if (!passContinues) {
      this.emit("revealPassCompleted");
      // The caller (BunkerRoom) decides when to actually move to Discussion,
      // via startDiscussionPhase() below.
    }
  }

  // Simplified form (no double-negative), same correction already applied
  // during the C# version's own section 1.4.
  hasCompletedRevealPass(): boolean {
    return this.activePlayers().every((p) => p.getRevealedCategories().size > 0);
  }

  // --- Special cards -----------------------------------------------------

  // General-purpose entry point for effects that don't need extra parameters
  // (VoteImmunity, RevealHiddenTrait, ForceRevealAll).
  useSpecialCard(casterPlayerId: string, targetPlayerId: string | null): SpecialCardEffectResult {
    const caster = this.getPlayer(casterPlayerId);
    if (!caster || caster.isEliminated) return fail("Invalid caster.");
    if (caster.hasUsedSpecialCard) return fail("Special card already used.");
    if (!caster.special) return fail("Player has no special card.");

    const effect = this.effectRegistry.tryGet(caster.special.effectType);
    if (!effect) return fail(`Unknown effect type: ${caster.special.effectType}`);

    const target = targetPlayerId ? this.getPlayer(targetPlayerId) ?? null : null;

    if (!effect.canApply(this, caster, target)) return fail("Effect cannot be applied in this context.");

    let result: SpecialCardEffectResult;
    switch (caster.special.effectType) {
      case SpecialCardEffectType.RevealHiddenTrait:
        result = this.applyRevealHiddenTrait(caster, target!);
        break;
      case SpecialCardEffectType.VoteImmunity:
        result = this.applyVoteImmunity(effect, caster, target);
        break;
      case SpecialCardEffectType.ForceRevealAll:
        result = this.applyForceRevealAll(caster, target!);
        break;
      default:
        result = fail("Effect requires extra parameters, use the dedicated method.");
    }

    if (result.success) {
      caster.hasUsedSpecialCard = true;
      this.emit("specialCardUsed", caster, caster.special);
    }

    return result;
  }

  // Dedicated method for SwapTrait, which needs to know which category to swap.
  useSwapTraitSpecialCard(
    casterPlayerId: string,
    targetPlayerId: string,
    category: CardCategory
  ): SpecialCardEffectResult {
    const caster = this.getPlayer(casterPlayerId);
    const target = this.getPlayer(targetPlayerId);

    if (!caster || !target || caster.isEliminated || target.isEliminated) {
      return fail("Invalid caster or target.");
    }
    if (caster.hasUsedSpecialCard) return fail("Special card already used.");
    if (caster.special?.effectType !== SpecialCardEffectType.SwapTrait) {
      return fail("Player's special card is not a swap effect.");
    }
    if (caster.isCategoryRevealed(category) || target.isCategoryRevealed(category)) {
      return fail("Category must be hidden for both players.");
    }

    const casterTrait = caster.getTrait(category)!;
    const targetTrait = target.getTrait(category)!;

    caster.replaceTrait(category, targetTrait);
    target.replaceTrait(category, casterTrait);

    caster.hasUsedSpecialCard = true;
    this.emit("specialCardUsed", caster, caster.special);

    return ok();
  }

  private applyRevealHiddenTrait(caster: PlayerData, target: PlayerData): SpecialCardEffectResult {
    const hiddenTrait = target.getTraits().find((t) => !target.isCategoryRevealed(t.category));
    if (!hiddenTrait) return fail("Target has no hidden traits.");

    // Does NOT call target.revealCategory() - stays hidden from everyone
    // else. Only the caster receives this trait in the result, delivered
    // privately by BunkerRoom (not broadcast via traitRevealed).
    return ok(hiddenTrait);
  }

  private applyVoteImmunity(
    effect: ISpecialCardEffect,
    caster: PlayerData,
    target: PlayerData | null
  ): SpecialCardEffectResult {
    effect.apply(this, caster, target);
    return ok();
  }

  private applyForceRevealAll(caster: PlayerData, target: PlayerData): SpecialCardEffectResult {
    const hiddenTrait = target.getTraits().find((t) => !target.isCategoryRevealed(t.category));
    if (!hiddenTrait) return fail("Target has no hidden traits.");

    target.revealCategory(hiddenTrait.category);
    this.emit("traitRevealed", target, hiddenTrait); // public - everyone sees it

    return ok(hiddenTrait);
  }

  // --- Discussion phase -------------------------------------------------

  // Called once the reveal pass is done (see revealPassCompleted above).
  // durationSeconds is passed by the caller (BunkerRoom) since GameSession
  // doesn't own real time.
  startDiscussionPhase(durationSeconds: number): void {
    if (this.phase !== GamePhase.Reveal) {
      console.warn("[GameSession] startDiscussionPhase called outside of Reveal phase.");
      return;
    }

    this.setPhase(GamePhase.Discussion);
    this.emit("discussionStarted", durationSeconds);
  }

  // Called by the caller's timer once discussion time is up.
  startVotingPhase(): void {
    if (this.phase !== GamePhase.Discussion) {
      console.warn("[GameSession] startVotingPhase called outside of Discussion phase.");
      return;
    }

    this.tiebreakerCandidateIds = null;
    this.tiebreakerAttempts = 0;
    this.currentVotes.clear();
    this.setPhase(GamePhase.Voting);
  }

  // --- Voting ------------------------------------------------------------

  castVote(voterPlayerId: string, targetPlayerId: string): boolean {
    const inTiebreaker = this.phase === GamePhase.VotingTiebreaker;

    if (this.phase !== GamePhase.Voting && !inTiebreaker) {
      console.warn("[GameSession] Vote cast outside of Voting phase, ignored.");
      return false;
    }

    const voter = this.getPlayer(voterPlayerId);
    if (!voter || voter.isEliminated) {
      console.warn(`[GameSession] Invalid voter: ${voterPlayerId}`);
      return false;
    }

    if (voterPlayerId === targetPlayerId) {
      console.warn("[GameSession] Self-voting is not allowed.");
      return false;
    }

    const target = this.getPlayer(targetPlayerId);
    if (!target || target.isEliminated) {
      console.warn(`[GameSession] Invalid vote target: ${targetPlayerId}`);
      return false;
    }

    if (target.hasVoteImmunityThisRound) {
      console.warn(`[GameSession] ${targetPlayerId} is immune this round, vote rejected.`);
      return false;
    }

    // During a tiebreaker, votes may only go to the tied candidates from
    // the previous round of voting.
    if (inTiebreaker && !this.tiebreakerCandidateIds!.includes(targetPlayerId)) {
      console.warn("[GameSession] Vote target is not part of the tiebreaker candidates.");
      return false;
    }

    this.currentVotes.set(voterPlayerId, targetPlayerId);
    return true;
  }

  // Called by the caller once every active player has voted (or a voting
  // timer expired). Resolves the round and transitions phases accordingly.
  resolveVotes(): VotingResult {
    const result = this.tallyVotes();
    this.emit("votingResolved", result);

    switch (result.resultType) {
      case VotingOutcome.PlayerEliminated:
        result.eliminatedPlayer!.isEliminated = true;
        this.emit("playerEliminated", result.eliminatedPlayer!);
        this.finishRound();
        break;

      case VotingOutcome.TieRequiresRevote:
        this.tiebreakerCandidateIds = result.tiedCandidates!.map((p) => p.playerId);
        this.tiebreakerAttempts++;
        this.currentVotes.clear();
        this.setPhase(GamePhase.VotingTiebreaker);
        break;

      case VotingOutcome.TieUnresolvedNoElimination:
      case VotingOutcome.NoVotesCast:
        this.finishRound();
        break;
    }

    return result;
  }

  private tallyVotes(): VotingResult {
    const voteCounts = new Map<string, number>();

    for (const targetId of this.currentVotes.values()) {
      voteCounts.set(targetId, (voteCounts.get(targetId) ?? 0) + 1);
    }

    if (voteCounts.size === 0) {
      return {
        resultType: VotingOutcome.NoVotesCast,
        eliminatedPlayer: null,
        tiedCandidates: null,
        voteCounts,
      };
    }

    const topCount = Math.max(...voteCounts.values());
    const topIds = [...voteCounts.entries()].filter(([, count]) => count === topCount).map(([id]) => id);

    if (topIds.length === 1) {
      return {
        resultType: VotingOutcome.PlayerEliminated,
        eliminatedPlayer: this.getPlayer(topIds[0]) ?? null,
        tiedCandidates: null,
        voteCounts,
      };
    }

    // Tie: decide whether to allow a re-vote or give up.
    const tiedPlayers = topIds.map((id) => this.getPlayer(id)!).filter(Boolean);
    const canRetry = this.tiebreakerAttempts < GameSession.MAX_TIEBREAKER_ATTEMPTS;

    return {
      resultType: canRetry ? VotingOutcome.TieRequiresRevote : VotingOutcome.TieUnresolvedNoElimination,
      eliminatedPlayer: null,
      tiedCandidates: tiedPlayers,
      voteCounts,
    };
  }

  // --- Round wrap-up -------------------------------------------------

  private finishRound(): void {
    this.setPhase(GamePhase.RoundResult);

    if (this.isGameOver()) {
      this.endGame();
    } else {
      this.startNextRound();
    }
  }

  // --- Win condition -----------------------------------------------------

  protected isGameOver(): boolean {
    return this.activePlayers().length <= this.config.survivorsTarget;
  }

  protected endGame(): void {
    this.setPhase(GamePhase.GameOver);

    const survivors = this.activePlayers();
    const reason =
      survivors.length > 0 ? GameOverReason.SurvivorsTargetReached : GameOverReason.AllPlayersEliminated;

    this.emit("gameOverResolved", { endReason: reason, survivors });
  }

  // --- Helpers ---------------------------------------------------------

  activePlayers(): PlayerData[] {
    return this.players.filter((p) => !p.isEliminated);
  }

  getPlayer(playerId: string): PlayerData | undefined {
    return this.players.find((p) => p.playerId === playerId);
  }
}
